homeassistant/components/qnap_qsw/config_flow.py

```diff
@@ -0,0 +1,65 @@
+"""Config flow for QNAP QSW."""
+from __future__ import annotations
+
+from typing import Any
+
+from aioqsw.exceptions import LoginError, QswError
+from aioqsw.localapi import ConnectionOptions, QnapQswApi
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_URL, CONF_USERNAME
+from homeassistant.data_entry_flow import AbortFlow, FlowResult
+from homeassistant.helpers import aiohttp_client, config_validation as cv
+from homeassistant.helpers.device_registry import format_mac
+
+from .const import DOMAIN
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle config flow for a QNAP QSW device."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors = {}
+
+        if user_input is not None:
+            url = user_input[CONF_URL]
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+
+            qsw = QnapQswApi(
+                aiohttp_client.async_get_clientsession(self.hass),
+                ConnectionOptions(url, username, password),
+            )
+
+            try:
+                system_board = await qsw.validate()
+            except LoginError:
+                errors[CONF_PASSWORD] = "invalid_auth"
+            except QswError:
+                errors[CONF_URL] = "cannot_connect"
+            else:
+                mac = system_board.get_mac()
+                if mac is None:
+                    raise AbortFlow("invalid_id")
+
+                await self.async_set_unique_id(format_mac(mac))
+                self._abort_if_unique_id_configured()
+
+                title = f"QNAP {system_board.get_product()} {mac}"
+                return self.async_create_entry(title=title, data=user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_URL): cv.url_no_path,
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
```

Comment:
I'm not sure if these will work since these get sent to the frontend with https://github.com/home-assistant-libs/voluptuous-serialize

Reply:
```
2022-04-22 19:24:29 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/ban.py", line 79, in ban_middleware
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/auth.py", line 219, in auth_middleware
    return await handler(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/view.py", line 137, in handle
    result = await result
  File "/Users/bdraco/home-assistant/homeassistant/components/config/config_entries.py", line 174, in post
    return await super().post(request)
  File "/Users/bdraco/home-assistant/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/Users/bdraco/home-assistant/homeassistant/helpers/data_entry_flow.py", line 85, in post
    result = self._prepare_result_json(result)
  File "/Users/bdraco/home-assistant/homeassistant/components/config/config_entries.py", line 183, in _prepare_result_json
    return _prepare_config_flow_result_json(result, super()._prepare_result_json)
  File "/Users/bdraco/home-assistant/homeassistant/components/config/config_entries.py", line 147, in _prepare_config_flow_result_json
    return prepare_result_json(result)
  File "/Users/bdraco/home-assistant/homeassistant/helpers/data_entry_flow.py", line 44, in _prepare_result_json
    data["data_schema"] = voluptuous_serialize.convert(
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/voluptuous_serialize/__init__.py", line 40, in convert
    pval = convert(value, custom_serializer=custom_serializer)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/voluptuous_serialize/__init__.py", line 113, in convert
    if issubclass(schema, Enum):
TypeError: issubclass() arg 1 must be a class
```

Reply:
Reverted in ef00b05a5b0b37d6d7cb394f23cbf13680203a58
Sorry!